include/dali/core/bitmask.h

```diff
@@ -0,0 +1,231 @@
+    if (start >= end)
+      return;
+    ptrdiff_t start_idx = start >> storage_bits_log;
+    ptrdiff_t end_idx = end >> storage_bits_log;
+    bit_storage_t ones = ~bit_storage_t(0);
+    bit_storage_t start_mask = ones;
+    bit_storage_t end_mask = 0;
+    if (start & (storage_bits - 1)) {
+        start_mask <<= (start & (storage_bits - 1));
+    }
+    if (end & (storage_bits - 1)) {
+        end_mask = ones >> (storage_bits - (end & (storage_bits - 1)));
+    }
+    if (start_idx == end_idx) {
+        start_mask &= end_mask;
+    }
+    ptrdiff_t idx = start_idx;
+    if (value)
+        storage_[idx++] |= start_mask;
+    else
+        storage_[idx++] &= ~start_mask;
+    if (value) {
+        for (; idx < end_idx; idx++)
+            storage_[idx] = ones;
+    } else {
+        for (; idx < end_idx; idx++)
+            storage_[idx] = 0;
+    }
+    if (end_mask && start_idx != end_idx) {
+      if (value)
+        storage_[end_idx] |= end_mask;
+      else
+        storage_[end_idx] &= ~end_mask;
+    }
+  }
+
+  /**
+   * @brief Fill the entire mask with given value.
+   */
+  void fill(bool value) {
+    size_t s = size();
+    clear();
+    resize(s, value);
+  }
+
+  /**
+   * @brief A pseudoreference to a single bit in the mask
+   */
+  struct bitref {
+    bitref &operator=(bool value) {
+      if (value)
+        *storage |= mask;
+      else
+        *storage &= ~mask;
+      return *this;
+    }
+    constexpr operator bool() const {
+      return *storage & mask;
+    }
+   private:
+    friend class bitmask;
+    constexpr bitref(bit_storage_t *array, ptrdiff_t index)
+    : storage(&array[index >> storage_bits_log])
+    , mask(bit_storage_t(1) << (index & (storage_bits - 1))) {}
+
+    bit_storage_t *storage;
+    bit_storage_t mask;
+  };
+
+  using reference_type = bitref;
+  using value_type = bool;
+
+  void push_back(bool value) {
+    size_t req_words = (size_ + storage_bits) >> storage_bits_log;
+    if (req_words > storage_.size())
+      storage_.push_back(value ? 1 : 0);  // put LSB in the storage vector
+    else
+      (*this)[size_] = value;  // set the last bit
+    size_++;
+  }
+
+  void pop_back() {
+    assert(!empty());
+    (*this)[--size_] = 0;
+    if ((size_ & (storage_bits - 1)) == 0)
+      storage_.pop_back();
+  }
+
+  constexpr bool empty() const noexcept { return size_ == 0; }
+
+  void reserve(size_t capacity) {
+    storage_.reserve((capacity + storage_bits - 1) >> storage_bits_log);
+  }
+
+  size_t capacity() const noexcept {
+    return storage_.capacity() * storage_bits;
+  }
+
+  void resize(size_t new_size, bool value = false) {
+    if (new_size == size())
+      return;
+    bit_storage_t ones = ~bit_storage_t(0);
+    bit_storage_t fill = value ? ones : 0;
+    size_t prev_words = storage_.size();
+    storage_.resize((new_size + storage_bits - 1) >> storage_bits_log, fill);
+    if (storage_.size() == prev_words && new_size > size()) {
```

Comment:
Is it possible that `storage_.size()` before and after resize is the same?

Reply:
It is - when resizing, say, from 13 to 17 bits - it's still just one word.